### ISSUE TITLE
Weighted probability?

### ISSUE THREAD

**Issue by Boarders** [jtobin/mwc-probability#15]
In the functions `multinomial` and `categorical` you have the condition that:

```Note that ps is a vector of probabilities and should sum to one.```

This is quite inconvenient† and I don't think it is strictly necessary. The code for categorical looks as follows:
```haskell
categorical :: (Foldable f, PrimMonad m) => f Double -> Prob m Int
categorical ps = do
  xs <- multinomial 1 ps
  case xs of
    [x] -> return x
    _   -> error "mwc-probability: invalid probability vector"
```
The code in multinomial in turn looks as follows:
```haskell
multinomial :: (Foldable f, PrimMonad m) => Int -> f Double -> Prob m [Int]
multinomial n ps = do
  let cumulative = scanl1 (+) (F.toList ps)
  replicateM n $ do
    z <- uniform
    case findIndex (> z) cumulative of
      Just g  -> return g
      Nothing -> error "mwc-probability: invalid probability vector"
```
We could change this to be as follows:
```haskell
multinomial :: (Foldable f, PrimMonad m) => Int -> f Double -> Prob m [Int]
multinomial n ps = do
  let (runningSum, total) = runningSums (F.toList ps)
  replicateM n $ do
    z <- uniformR (0,total)
    case findIndex (> z) cumulative of
      Just g  -> return g
      Nothing -> error "mwc-probability: invalid probability vector"

runningSums :: Num a => [a] -> ([a], a)
runningSums = let adds = scanl1 (+) xs in (sum xs, adds)

-- Note: GHC does well at optimising this and it performs significantly
-- better than writing a single pass recursion by hand.
```

Would you be open to a pull request with this change?

† In particular it is often the case that a weighted probability distribution is more natural. For instance I was wanting to write a kmeans++ algorithm and this gets the probabilities proportional to the squared distance from points you have selected so far. With the API as it currently is I am forced to fmap dividing by the total which is not very efficient.

**Comment by jtobin** [comment]
Yup, I think you're right about this.  Feel free to make a PR!

**Comment by jtobin** [comment]
Resolved in #16.
